sdk/nodejs/runtime/state.ts

```diff
@@ -0,0 +1,150 @@
+// Copyright 2016-2018, Pulumi Corporation.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import { AsyncLocalStorage } from "async_hooks";
+import { Stack } from "./stack";
+import * as config from "./config";
+
+const nodeEnvKeys = {
+    project: "PULUMI_NODEJS_PROJECT",
+    stack: "PULUMI_NODEJS_STACK",
+    dryRun: "PULUMI_NODEJS_DRY_RUN",
+    queryMode: "PULUMI_NODEJS_QUERY_MODE",
+    parallel: "PULUMI_NODEJS_PARALLEL",
+    monitorAddr: "PULUMI_NODEJS_MONITOR",
+    engineAddr: "PULUMI_NODEJS_ENGINE",
+    syncDir: "PULUMI_NODEJS_SYNC",
+    // this value is not set by the CLI and is controlled via a user set env var unlike the values above
+    cacheDynamicProviders: "PULUMI_NODEJS_CACHE_DYNAMIC_PROVIDERS",
+    organization: "PULUMI_NODEJS_ORGANIZATION",
+};
+
+const pulumiEnvKeys = {
+    legacyApply: "PULUMI_ENABLE_LEGACY_APPLY",
+};
+
+export const asyncLocalStorage = new AsyncLocalStorage();
```

Comment:
I believe you can add type information:

```suggestion
export const asyncLocalStorage = new AsyncLocalStorage<Store>();
```